File: stack/stack.js
/**
 * |       |
 * |   5   |  -> Top
 * |   9   |
 * |   2   |
 * |   4   |
 * |_______|
 *
 */

function Stack() {
    let stackArray = [];
    const publicAPI = {
        top() {
            if (this.isEmpty()) {
                console.log("Stack is empty");
                return;
            }
            return stackArray[stackArray.length - 1];
        },
        push(value) {
            stackArray.push(value);
        },
        pop() {
            if (this.isEmpty()) {
                console.log("Stack is empty");
                return;
            }
            const poppedElement = stackArray.pop();
            return poppedElement;
        },
        isEmpty() {
            return stackArray.length === 0;
        },
    };
    return publicAPI;
}

// class Stack {
//     constructor() {
//         this.stackArray = [];
//     }
//     top() {
//         if (this.isEmpty()) {
//             console.log("Stack is empty");
//             return;
//         }
//         return this.stackArray[this.stackArray.length - 1];
//     }
//     push(value) {
//         this.stackArray.push(value);
//     }
//     pop() {
//         if (this.isEmpty()) {
//             console.log("Stack is empty");
//             return;
//         }
//         const poppedElement = this.stackArray.pop();
//         return poppedElement;
//     }
//     isEmpty() {
//         return this.stackArray.length === 0;
//     }
// }

const stack = new Stack();

console.log(stack.isEmpty());

stack.push(3);
stack.push(4);
stack.push(5);

stack.pop();

console.log(stack.top());
console.log(stack.isEmpty());

File: dsainjava/stack/Stack.java
package stack;

public class Stack {
    private int top = -1, size;
    private int[] arr;

    public Stack(int size) {
        this.size = size;
        this.arr = new int[size];
    }

    public boolean overflow() {
        return top == size - 1;
    }

    public boolean underflow() {
        return top == -1;
    }

    public void push(int x) {
        if (overflow()) {
            System.out.println("Stack is full");
        } else {
            top++;
            System.out.println("top: " + top);
            arr[top] = x;
        }
    }

    public void pop() {
        if (underflow()) {
            System.out.println("Stack is empty");
        } else {
            top--;
        }
    }

    public int peek() {
        if (top != -1)
            return -1;
        return arr[top];
    }

    public void display() {
        for (int i = top; i > -1; i--) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}
